matrix_sdk_appservice/src/lib.rs

```diff
@@ -214,66 +213,122 @@ impl Appservice {
+            ClientConfig::default(),
+        )
+        .await?;
+
+        Ok(appservice)
+    }
+
+    /// Same as [`Self::new()`] but lets you provide a [`ClientConfig`] for the
+    /// [`Client`]
+    pub async fn new_with_client_config(
+        homeserver_url: impl TryInto<Url, Error = url::ParseError>,
+        server_name: impl TryInto<ServerNameBox, Error = identifiers::Error>,
+        registration: AppserviceRegistration,
+        client_config: ClientConfig,
     ) -> Result<Self> {
         let homeserver_url = homeserver_url.try_into()?;
         let server_name = server_name.try_into()?;
+        let registration = Arc::new(registration);
+        let clients = Arc::new(DashMap::new());
 
-        let client_sender_localpart = Client::new(homeserver_url.clone())?;
+        let mut appservice = Appservice { homeserver_url, server_name, registration, clients };
 
-        client_session_with_login_restore(
-            &client_sender_localpart,
-            &registration,
-            registration.sender_localpart.as_ref(),
-            &server_name,
-        )
-        .await?;
+        // we cache the [`MainAppserviceUser`] by default
+        appservice.client_with_config(None, client_config).await?;
 
-        Ok(Appservice { homeserver_url, server_name, registration, client_sender_localpart })
+        Ok(appservice)
     }
 
-    /// Get a [`Client`]
+    /// Create a [`Client`]
+    ///
+    /// Will create and return a [`Client`] that's configured to [assert the
+    /// identity] on all outgoing homeserver requests if `localpart` is
+    /// given. If not given the [`Client`] will use the [`MainAppserviceUser`].
     ///
-    /// Will return a `Client` that's configured to [assert the identity] on all
-    /// outgoing homeserver requests if `localpart` is given. If not given
-    /// the `Client` will use the main user associated with this appservice,
-    /// that is the `sender_localpart` in the [`AppserviceRegistration`]
+    /// This method is a singleton that saves the client internally for re-use
+    /// based on the `localpart`.
     ///
     /// # Arguments
     ///
     /// * `localpart` - The localpart of the user we want assert our identity to
     ///
     /// [registration]: https://matrix.org/docs/spec/application_service/r0.1.2#registration
     /// [assert the identity]: https://matrix.org/docs/spec/application_service/r0.1.2#identity-assertion
-    pub async fn client(&self, localpart: Option<&str>) -> Result<Client> {
+    pub async fn client(&mut self, localpart: Option<&str>) -> Result<Client> {
+        let client = self.client_with_config(localpart, ClientConfig::default()).await?;
+
+        Ok(client)
+    }
+
+    /// Same as [`Self::client`] but with the ability to pass in a
+    /// [`ClientConfig`]
+    ///
+    /// Since this method is a singleton follow-up calls with different
+    /// [`ClientConfig`]s will be ignored.
+    pub async fn client_with_config(
+        &mut self,
```

Comment:
Is that `&mut` still needed? I don't think you need it since the clients are in the dashmap.

Reply:
Good catch, isn't needed anymore indeed

I wonder what it takes to teach clippy about DashMap :thinking: